api/lib/subscription/slashing.js

```diff
@@ -38,20 +63,69 @@ class SlashingMonitor {
     )
 
     // requires some more logic to not spam the notifications if a validator is down for 1000 blocks
-    // this.client.subscribe(
-    //   { query: 'liveness.missed_blocks >= 1' },
-    //   (response) => {
-    //     const missedBlocks = response.events['liveness.address'].map(
-    //       (address, index) => ({
-    //         operatorAddress: address,
-    //         missedBlocks: response.events['liveness.missed_blocks'][index],
-    //         height: response.height
-    //       })
-    //     )
-    //     console.log('Missed block', missedBlocks)
-    //   }
-    // )
+    this.client.subscribe(
+      { query: `tm.event='NewBlockHeader' AND liveness.missed_blocks >= 1` },
+      async (response) => {
+        try {
+          const missedBlocks = response.events['liveness.address'].map(
+            (address, index) => ({
+              networkId: this.networkId,
+              operatorAddress: address,
+              blocks: response.events['liveness.missed_blocks'][index],
+              reason: 'missed_blocks',
+              height: response.data.value.header.height
+            })
+          )
+          const rows = await Promise.all(
+            missedBlocks.map(async (missedBlockEvent) => {
+              const lastMissedBlockEvent = await getLastMissedBlock(
+                this.networkId,
+                missedBlockEvent.operatorAddress
+              )
+              return aggregateMissedBlocks(
+                lastMissedBlockEvent,
+                missedBlockEvent
+              )
+            })
+          )
+          database(config)('').upsert('slashes', rows)
+          console.log('Wrote missed block events', rows.length)
+        } catch (error) {
+          console.error('Failed to write missed block events', error)
+          Sentry.captureException(error)
+        }
+      }
+    )
+  }
+}
+
+async function getLastMissedBlock(networkId, operatorAddress) {
+  // get last liveness issue
+  const { data } = await database(config)('').query(`
+    query {
+      slashes(where:{networkId:{_eq:"${networkId}"}, operatorAddress:{_eq:"${operatorAddress}"}, reason:{_eq:"missed_blocks"}}, order_by:{id:desc}, limit: 1) {
+        height
+        id
+        reason
+      }
+    }
+  `)
```

Comment:
Why not add a `liveness` query too? Or `missedBlocks`. Since liveness and slashes are two different things. Could be very useful for validators this query. Could look like this:

```
    query {
      missedBlocks(where:{networkId:{_eq:"${networkId}"}, operatorAddress:{_eq:"${operatorAddress}"}, order_by:{height:desc}) {
        height
        id # from Block
        total # total number from blocks missed. Could be a separate query by itself
      }

```

We can add this in a separate PR though

Reply:
I like it. Can you open an issue? We might need this for the validator profiles (uptime per chain and cross chain).

Reply:
> We could fire an event only on "validator goes offline for at least one block" and "validator got slashed and jailed for downtime".

You mean as two separate events or you mean handling both as the same event?

Reply:
those would be 2 different events. one on the first block, one on the 5000 block or sth.